pkg/reconciler/v1alpha1/configuration/configuration.go

```diff
@@ -230,3 +254,59 @@ func (c *Reconciler) updateStatus(u *v1alpha1.Configuration) (*v1alpha1.Configur
 	}
 	return newu, nil
 }
+
+func (c *Reconciler) gcRevisions(ctx context.Context, config *v1alpha1.Configuration) error {
+	logger := logging.FromContext(ctx)
+
+	selector := labels.Set{serving.ConfigurationLabelKey: config.Name}.AsSelector()
+	revs, err := c.revisionLister.Revisions(config.Namespace).List(selector)
+	if err != nil {
+		return err
+	}
+
+	for _, rev := range revs {
+		gcConfig := configns.FromContext(ctx).RevisionGc
+
+		if isStale, err := isRevisionStale(gcConfig, rev, config.Spec.Generation); err != nil {
+			logger.Errorf("Failed to determine staleness of revision: %v", err)
+			continue
+		} else if isStale {
+			lastPinStr, _ := rev.ObjectMeta.Annotations[serving.RevisionLastPinnedAnnotationKey]
+			logger.Infof("Removing stale revision %v with creation time %v and lastPinned time %v.", rev.ObjectMeta.Name, rev.ObjectMeta.CreationTimestamp, lastPinStr)
+			err := c.ServingClientSet.ServingV1alpha1().Revisions(rev.Namespace).Delete(rev.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				logger.Errorf("Failed to delete stale revision: %v", err)
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func isRevisionStale(cfg *gc.Config, rev *v1alpha1.Revision, curGeneration int64) (bool, error) {
+	// maxGen is the maximum generation number we consider for GC
+	maxGen := curGeneration - cfg.StaleRevisionMinimumGenerations
+
+	// Check if rev is within "MinimumGenerations" of latest
+	if gen, err := rev.GetConfigurationGeneration(); err != nil {
+		return false, err
+	} else if gen > maxGen {
+		return false, nil
+	}
+
+	curTime := time.Now()
+	if rev.ObjectMeta.CreationTimestamp.Add(cfg.StaleRevisionCreateDelay).After(curTime) {
+		// Revision was created sooner than staleRevisionCreateDelay. Ignore it
+		return false, nil
+	}
+
+	lastPin, err := rev.GetLastPinned()
+	if err != nil && err.(v1alpha1.LastPinnedParseError).Type == v1alpha1.AnnotationParseErrorTypeInvalid {
+		return false, err
+	}
+
```

Comment:
Should we be handling the `AnnotationParseErrorTypeMissing` error here? As written, if the annotation is missing then we ignore that error and the `lastPin` value will end up being Jan 1st 0001 which will get marked as stale and thus the revision garbage collected, assuming the minimum generations condition is met above.

I know we're still in alpha and don't explicitly support upgrading, but in the case of an upgrade from a prior Knative we'll definitely have a lot of missing lastPinned values. Or in the case of downtime/outage of the controller pod longer than the `StaleRevisionCreateDelay` we could also end up in a situation with missing lastPinned values that potentially lead to unintended GC here?

It does seem like lastPinned would get set the next time the Route reconciles, but with the recently increased resync periods that could be a really long time. And there's no guarantee the Route reconciler will add a lastPinned before the Configuration reconciler checks for staleness?

Reply:
Another case we should think about, since this uses wall clock times, is what happens if I'm working locally in minikube, deploy a bunch of stuff, put my laptop to sleep, and come back over the weekend. When everything comes back up will all my revisions potentially get garbage collected?

Reply:
The logic is a bit wonky here so I think I'm going to change it around to explicitly handle the missing case as @mattmoor has a similar question. As is though, if we hit the missing annotation case we fail the `if lastPin.Add(cfg.StaleRevisionTimeout).Before(curTime)` check below and therefore `return false, nil` which seems appropriate - missing annotation is not an error we want to propogate but the revision is not considered stale. This should also work fine for upgrade where annotations are missing.

For a case where we get a long sleep youre correct that the GC will potentially get greedy when it wakes up. I'm not sure theres much we can do for this case beyond advise folks using a setup where clocks can jump significantly to use the minimum-stale-revisions config option or disable GC.

Reply:
I take that back - youre totally right that in the missing case this is going to return that the revision is stale. I'll update this and add a test :)

Reply:
Here's the Go Playground snippet I was using to test that logic - https://play.golang.org/p/BN09jiGwhDU

Reply:
Updated this so a missing label is treated as non stale which should fix the initial and upgrade cases.